? icewm-alpha-v1.patch
? icewm-alpha-v2.patch
? icewm-alpha-v3.patch
? icewm-dbg-v1.patch
? src/tmp.txt
Index: src/wmcontainer.cc
===================================================================
RCS file: /cvsroot/icewm/icewm-1.2/src/wmcontainer.cc,v
retrieving revision 1.18
diff -u -r1.18 wmcontainer.cc
--- src/wmcontainer.cc	16 Apr 2006 16:28:40 -0000	1.18
+++ src/wmcontainer.cc	20 Jul 2006 16:19:28 -0000
@@ -14,8 +14,9 @@
 
 #include <stdio.h>
 
-YClientContainer::YClientContainer(YWindow *parent, YFrameWindow *frame)
-:YWindow(parent)
+YClientContainer::YClientContainer(YWindow *parent, YFrameWindow *frame,
+                                   Visual *visual)
+    :YWindow(parent, 0, visual)
 {
     fFrame = frame;
     fHaveGrab = false;
Index: src/wmcontainer.h
===================================================================
RCS file: /cvsroot/icewm/icewm-1.2/src/wmcontainer.h,v
retrieving revision 1.2
diff -u -r1.2 wmcontainer.h
--- src/wmcontainer.h	17 May 2003 17:41:25 -0000	1.2
+++ src/wmcontainer.h	20 Jul 2006 16:19:28 -0000
@@ -7,7 +7,7 @@
 
 class YClientContainer: public YWindow {
 public:
-    YClientContainer(YWindow *parent, YFrameWindow *frame);
+    YClientContainer(YWindow *parent, YFrameWindow *frame, Visual *visual);
     virtual ~YClientContainer();
 
     virtual void handleButton(const XButtonEvent &button);
Index: src/wmframe.cc
===================================================================
RCS file: /cvsroot/icewm/icewm-1.2/src/wmframe.cc,v
retrieving revision 1.96
diff -u -r1.96 wmframe.cc
--- src/wmframe.cc	16 Apr 2006 16:28:40 -0000	1.96
+++ src/wmframe.cc	20 Jul 2006 16:19:29 -0000
@@ -48,7 +48,8 @@
     return false;
 }
 
-YFrameWindow::YFrameWindow(YWindow *parent): YWindow(parent) {
+YFrameWindow::YFrameWindow(YWindow *parent, Visual *visual)
+    : YWindow(parent, 0, visual) {
     if (activeBorderBg == 0)
         activeBorderBg = new YColor(clrActiveBorder);
     if (inactiveBorderBg == 0)
@@ -126,7 +127,7 @@
 
     createPointerWindows();
 
-    fClientContainer = new YClientContainer(this, this);
+    fClientContainer = new YClientContainer(this, this, visual);
     fClientContainer->show();
 
     fTitleBar = new YFrameTitleBar(this, this);
Index: src/wmframe.h
===================================================================
RCS file: /cvsroot/icewm/icewm-1.2/src/wmframe.h,v
retrieving revision 1.30
diff -u -r1.30 wmframe.h
--- src/wmframe.h	16 Apr 2006 16:28:40 -0000	1.30
+++ src/wmframe.h	20 Jul 2006 16:19:29 -0000
@@ -20,7 +20,7 @@
 
 class YFrameWindow: public YWindow, public YActionListener, public YTimerListener, public YPopDownListener, public YMsgBoxListener, public ClientData {
 public:
-    YFrameWindow(YWindow *parent);
+    YFrameWindow(YWindow *parent, Visual *visual = 0);
     virtual ~YFrameWindow();
 
     void doManage(YFrameClient *client);
Index: src/wmmgr.cc
===================================================================
RCS file: /cvsroot/icewm/icewm-1.2/src/wmmgr.cc,v
retrieving revision 1.119
diff -u -r1.119 wmmgr.cc
--- src/wmmgr.cc	16 Apr 2006 16:28:40 -0000	1.119
+++ src/wmmgr.cc	20 Jul 2006 16:19:30 -0000
@@ -1372,7 +1372,7 @@
 
     manager->updateFullscreenLayerEnable(false);
 
-    frame = new YFrameWindow(0);
+    frame = new YFrameWindow(0, client->visual());
     if (frame == 0) {
         delete client;
         goto end;
Index: src/ywindow.cc
===================================================================
RCS file: /cvsroot/icewm/icewm-1.2/src/ywindow.cc,v
retrieving revision 1.49
diff -u -r1.49 ywindow.cc
--- src/ywindow.cc	4 Feb 2006 16:46:27 -0000	1.49
+++ src/ywindow.cc	20 Jul 2006 16:19:31 -0000
@@ -18,6 +18,8 @@
 #include "ytimer.h"
 #include "ypopup.h"
 
+#include <X11/extensions/Xrender.h>
+
 /******************************************************************************/
 /******************************************************************************/
 
@@ -114,14 +116,14 @@
 /******************************************************************************/
 /******************************************************************************/
 
-YWindow::YWindow(YWindow *parent, Window win):
+YWindow::YWindow(YWindow *parent, Window win, Visual *visual):
     fParentWindow(parent),
     fNextWindow(0), fPrevWindow(0), fFirstWindow(0), fLastWindow(0),
     fFocusedWindow(0),
 
     fHandle(win), flags(0), fStyle(0), fX(0), fY(0), fWidth(1), fHeight(1),
     fPointer(), unmapCount(0),
-    fGraphics(0),
+    fGraphics(0), fVisual(visual),
     fEventMask(KeyPressMask|KeyReleaseMask|FocusChangeMask|
                LeaveWindowMask|EnterWindowMask),
     fWinGravity(NorthWestGravity), fBitGravity(ForgetGravity),
@@ -301,7 +303,7 @@
                                 0,
                                 CopyFromParent,
                                 (fStyle & wsInputOnly) ? InputOnly : InputOutput,
-                                CopyFromParent,
+                                (fVisual) ? fVisual : CopyFromParent,
                                 attrmask,
                                 &attributes);
 
@@ -311,6 +313,9 @@
 
         if ((flags & wfVisible) && !(flags & wfNullSize))
             XMapWindow(xapp->display(), fHandle);
+
+        // Zero visual so it is not used for system menu
+        fVisual = 0;
     } else {
         XWindowAttributes attributes;
 
@@ -352,6 +357,22 @@
         }
 
         XSelectInput(xapp->display(), fHandle, fEventMask);
+
+        // Get the visual if we have a window with alpha transparency
+        XWindowAttributes attr;
+        Status err = XGetWindowAttributes(xapp->display(), fHandle, &attr);
+        bool ok = (BadDrawable != err) && (BadWindow != err) && attr.visual;
+
+        XRenderPictFormat *format = NULL;
+        if (ok)
+            format = XRenderFindVisualFormat(xapp->display(), attr.visual);
+
+        if (format && format->type == PictTypeDirect &&
+            format->direct.alphaMask)
+        {
+            //printf("YWindow(0x%X) adopted alpha window\n", this);
+            fVisual = attr.visual;
+        }
     }
     XSaveContext(xapp->display(), fHandle, windowContext, (XPointer)this);
     flags |= wfCreated;
Index: src/ywindow.h
===================================================================
RCS file: /cvsroot/icewm/icewm-1.2/src/ywindow.h,v
retrieving revision 1.19
diff -u -r1.19 ywindow.h
--- src/ywindow.h	22 Jan 2006 13:01:19 -0000	1.19
+++ src/ywindow.h	20 Jul 2006 16:19:31 -0000
@@ -24,7 +24,7 @@
 
 class YWindow {
 public:
-    YWindow(YWindow *aParent = 0, Window win = 0);
+    YWindow(YWindow *aParent = 0, Window win = 0, Visual *visual = 0);
     virtual ~YWindow();
 
     void setStyle(unsigned long aStyle);
@@ -202,6 +202,8 @@
     bool hasPopup();
     void setDoubleBuffer(bool doubleBuffer);
 
+    Visual *visual() { return fVisual; }
+
 private:
     typedef enum {
         wfVisible   = 1 << 0,
@@ -234,6 +236,7 @@
     YCursor fPointer;
     int unmapCount;
     Graphics *fGraphics;
+    Visual *fVisual;
     long fEventMask;
     int fWinGravity, fBitGravity;
 
